contributors/design-proposals/code-generator/kubegen.md

```diff
@@ -0,0 +1,269 @@
+
+### Running the command line
+
+#### Use defaults
+
+```sh
+kubegen
+```
+
+- run code generators against discovered API type definitions
+- prepend license to generated files
+  - wrap the contents of the LICENSE in comments
+  - exit non-zero if the LICENSE file is missing
+
+#### Specifying copyright owners
+
+The copyright owners for the license can be defined with the `--copyright` flag
+
+```sh
+kubegen --copyright "The Kubernetes Authors."
+```
+
+- run code generators against discovered API type definitions
+- prepend license to generated files
+  - wrap the contents of the LICENSE in comments
+  - start license with "Copyright <current year> <copyright>" e.g. *Copyright 2018 The Kubernetes Authors.*
+  - exit non-zero if the LICENSE file is missing
+
+
+#### Specifying the license file
+
+The license file can be overridden from the LICENSE file using the `--license-file` flag
+
+```sh
+kubegen --license-file "boilerplate.go.txt"
+```
+
+- run code generators against discovered API type definitions
+- prepend license to generated files
+  - use the contents from the boilerplate.go.txt file instead of LICENSE
+  - if boilerplate.go.txt is already in comments then use it verbatim, otherwise wrap it in comments
+  - exit non-zero if boilerplate.go.txt is missing
+  
+**Note**: This passes the `--go-header-file` to each of the code generators
+
+#### Specifying a license type
+
+The license can be overridden from the LICENSE file using the `--license` flag
+
+```sh
+kubegen --license "Apache 2.0"
+```
+
+- run code generators against discovered API type definitions
+- prepend license to generated files
+  - use an *Apache 2.0* license
+
+#### Specifying external API versions
+
+Generating code only for specific API group versions can be performed by providing the API
+group versions as positional arguments.  When manually specifying the API group versions,
+neither internal nor external API types are automatically discovered.
+
+```sh
+kubegen apps/v1 apps/v1beta1 extensions/v1beta1
+```
+
+- run all code generators against the apps/v1 apps/v1beta1 extensions/v1beta1 API groups versions
+  - only run against external types, do not run against internal types
+- prepend license to generated files
+
+#### Specifying internal API groups
+
+Generating code only for specific API groups can be performed by providing the API
+groups as positional arguments.  When manually specifying the API groups,
+neither internal nor external API types are automatically discovered.
+
+```sh
+kubegen apps extensions
+```
+
+- run all code generators against the apps extensions API groups
+  - only run against internal types, do not run against external types
+- prepend license to generated files
+
+#### Specifying both internal API groups and external API versions
+
+To generate code for both manually specified internal and external types
+provide both as positional arguments.
+
+```sh
+kubegen apps apps/v1 apps/v1beta1 extensions extensions/v1beta1
```

Comment:
do you have to pass internal and external groups? Will conversions be created if you only pass the external groups? What about clientsets and informers?

Reply:
groups are always internal, external requires a version.